heron/packing/src/java/com/twitter/heron/packing/roundrobin/ResourceCompliantRRPacking.java

```diff
@@ -0,0 +1,419 @@
+   */
+  protected ArrayList<RamRequirement> getRAMInstances() {
+    ArrayList<RamRequirement> ramRequirements = new ArrayList<>();
+    Map<String, Integer> parallelismMap = TopologyUtils.getComponentParallelism(topology);
+    Map<String, Long> ramMap = TopologyUtils.getComponentRamMapConfig(topology);
+
+    for (String component : parallelismMap.keySet()) {
+      if (ramMap.containsKey(component)) {
+        if (!isValidInstance(ramMap.get(component), instanceCpuDefault, instanceDiskDefault)) {
+          return null;
+        } else {
+          ramRequirements.add(new RamRequirement(component, ramMap.get(component)));
+        }
+      } else {
+        if (!isValidInstance(instanceRamDefault, instanceCpuDefault, instanceDiskDefault)) {
+          return null;
+        } else {
+          ramRequirements.add(new RamRequirement(component, instanceRamDefault));
+        }
+      }
+    }
+    return ramRequirements;
+  }
+
+  /**
+   * Get the instances' allocation based on the Resource Compliant Round Robin algorithm
+   *
+   * @return Map &lt; containerId, list of InstanceId belonging to this container &gt;
+   */
+  protected Map<String, List<String>> getResourceCompliantRRAllocation() {
+
+    Map<String, List<String>> allocation = new HashMap<>();
+    ArrayList<Container> containers = new ArrayList<>();
+    ArrayList<RamRequirement> ramRequirements = getRAMInstances();
+
+    if (ramRequirements == null) {
+      return null;
+    }
+
+    int totalInstance = TopologyUtils.getTotalInstance(topology);
+
+    if (numContainer > totalInstance) {
+      throw new RuntimeException("More containers allocated than instance.");
+    }
+
+    for (int i = 1; i <= numContainer; ++i) {
+      allocation.put(getContainerId(i), new ArrayList<String>());
+    }
+    for (int i = 0; i <= numContainer - 1; i++) {
+      allocateNewContainer(containers);
+    }
+
+    int index = 1;
+    int globalTaskIndex = 1;
+    Map<String, Integer> parallelismMap = TopologyUtils.getComponentParallelism(topology);
+    int j = 0;
+    for (String component : parallelismMap.keySet()) {
+      long ramRequirement = ramRequirements.get(j).getRamRequirement();
+      int numInstance = parallelismMap.get(component);
+      for (int i = 0; i < numInstance; ++i) {
+        int containerId = index;
+        if (placeResourceCompliantRRInstance(containers, containerId,
+            ramRequirement, instanceCpuDefault, instanceDiskDefault)) {
+          allocation.get(getContainerId(containerId))
+              .add(getInstanceId(containerId, component, globalTaskIndex, i));
+
+        } else {
+          List<TopologyAPI.Config.KeyValue> topologyConfig
+              = topology.getTopologyConfig().getKvsList();
+          boolean adjustNumberOfContainers
+              = Boolean.parseBoolean(TopologyUtils.getConfigWithDefault(
+              topologyConfig, com.twitter.heron.api.Config.TOPOLOGY_ADJUST_NUMBER_CONTAINERS,
```

Comment:
@billonahill @kramasamy I can remove that but I thought it might be confusing for users to see a larger number of containers than what they declared. That is why I let them clarify the behavior. Do you think it is better to remove that and always adjust if needed?